private/pkg/git/git.go

```diff
@@ -352,6 +358,67 @@ func GetRefsForGitCommitAndRemote(
 	return refs, nil
 }
 
+// ReadFileAtRef will read the file at path rolled back to the given ref, if
+// it exists at that ref.
+//
+// Specifically, this will take path, and find the associated .git directory
+// (and thus, repository root) associated with that path, if any. It will then
+// look up the commit corresponding to ref in that git directory, and within
+// that commit, the blob corresponding to that path (relative to the repository
+// root).
+func ReadFileAtRef(
+	ctx context.Context,
+	runner command.Runner,
+	envContainer app.EnvContainer,
+	path, ref string,
+) ([]byte, error) {
+	orig := path
+	path, err := filepath.Abs(path)
+	if err != nil {
+		return nil, err
+	}
+
+	// Find a directory with a .git directory.
+	dir := filepath.Dir(path)
+	for {
+		_, err := os.Stat(filepath.Join(dir, ".git"))
+		if err == nil {
+			break
+		} else if errors.Is(err, os.ErrNotExist) {
+			parent := filepath.Dir(dir)
+			if parent == dir {
+				return nil, fmt.Errorf("could not find .git directory for %s: %w", orig, ErrNotARepo)
+			}
+			dir = parent
+		} else {
+			return nil, err
+		}
+	}
```

Comment:
We could use `CheckDirectoryIsValidGitCheckout` to check if this is a git repository:
https://github.com/bufbuild/buf/blob/b886118276b22562ead39d3debbffd1774460040/private/pkg/git/git.go#L215

The call-site will need to handle `ErrInvalidGitCheckout` instead.

Reply:
That's not quite what we want. We want to get the repository root, which requires traversing upwards (vs IsValidCheckout, which just kicks `rev-parse`).